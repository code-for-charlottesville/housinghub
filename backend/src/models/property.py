import json

from marshmallow_sqlalchemy import SQLAlchemyAutoSchema
from sqlalchemy import Column
from sqlalchemy.dialects.postgresql import (ARRAY, BOOLEAN, DATE, FLOAT,
                                            INTEGER, UUID, VARCHAR)

from models.base import Base


class Property(Base):
  __tablename__ = 'property'

  id = Column(UUID, primary_key = True)
  landlord_id = Column(UUID)
  navigator_id = Column(UUID)
  voucher_type_accepted = Column(ARRAY(VARCHAR))
  voucher_type_not_accepted = Column(ARRAY(VARCHAR))
  address = Column(VARCHAR)
  zip_code = Column(VARCHAR)
  unit_apt_no = Column(VARCHAR)
  property_name = Column(VARCHAR)
  bus_line = Column(BOOLEAN)
  school_district = Column(VARCHAR)
  wheelchair_accessibility = Column(BOOLEAN)
  elevator = Column(BOOLEAN)
  monthly_rent = Column(INTEGER)
  contact_method = Column(ARRAY(VARCHAR))
  is_available = Column(BOOLEAN)
  date_first_available = Column(DATE)
  last_contact_date = Column(DATE)
  potential_month_available = Column(INTEGER)
  bedrooms = Column(INTEGER)
  bathrooms = Column(INTEGER)
  shared_bathrooms = Column(INTEGER)
  has_basement = Column(BOOLEAN)
  application_fee = Column(FLOAT)
  deposit = Column(FLOAT)
  last_month_rent_required = Column(BOOLEAN)
  allow_criminal_records = Column(BOOLEAN)
  listing_date = Column(DATE)
  where_listed = Column(ARRAY(VARCHAR))
  floor = Column(INTEGER)
  housing_type = Column(VARCHAR)
  year_available = Column(INTEGER)
  credit_screening_company = Column(VARCHAR)
  background_screening_company = Column(VARCHAR)
  last_contacted_by = Column(UUID)
  pets_allowed = Column(BOOLEAN)
  background_check_required = Column(BOOLEAN)
  near_busstop = Column(BOOLEAN)
  primary_contact = Column(VARCHAR)
  contact_phone_number = Column(VARCHAR)

  def __repr__(self):
    _dict = PropertySchema().dump(self)
    return json.dumps(_dict)

class PropertySchema(SQLAlchemyAutoSchema):
  class Meta:
    model = Property
