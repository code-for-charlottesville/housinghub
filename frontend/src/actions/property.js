export const simpleAction = () => dispatch => {
 dispatch({
  type: 'SIMPLE_ACTION',
  payload: 'result_of_simple_action'
 })
}