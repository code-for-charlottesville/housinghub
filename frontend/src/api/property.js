import axios from "axios";
import apiEndpoint from "./endpoint";

/**
 * makes POST request to /backend/property with property
 **/

export function postProperty(property) {
  return axios
    .post(`${apiEndpoint}/property`, property)
    .then((r) => {
      console.log(property);
      return Promise.resolve(r.data);
    })
    .catch((err) => {
      console.log(err);
      return Promise.resolve(err.response.data);
    });
}
