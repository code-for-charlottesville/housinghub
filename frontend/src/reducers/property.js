export default (state = {}, action) => {
  switch (action.type) {
    case "SIMPLE_ACTION":
      return {
        result: action.payload
      };
    default:
      return state;
  }
};
