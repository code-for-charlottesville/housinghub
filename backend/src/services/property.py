import uuid
import app 
from sqlalchemy.orm import sessionmaker
from sqlalchemy import update
from models.property import Property


class PropertyService:

    def __init__(self, logger, Session):
        super().__init__()
        self.session = Session()
        self.db_session = Session()

    def add_property(self, property: Property):
        property.id = uuid.uuid4()
        self.session.add(property)
        self.session.commit()
        return property

    def get_property(self, payload):
        if (self.db_session.query(Property).all() == None):
            return None
        searchFields = payload["searchFields"]
        _properties = self.session.query(Property)
        for key, value in searchFields.items():
            if (isinstance(value, list)):
                # for houing types and zip code
                if (len(value) > 0):
                    _properties = _properties.filter(getattr(Property, key).in_(value))
                else:
                    continue
            elif (key == "date_available"):
                _properties = _properties.filter(Property.date_first_available <= value)
            elif (key == "max_rent"):
                _properties = _properties.filter(Property.monthly_rent <= value)
            else:
                _properties = _properties.filter(getattr(Property, key) == value)

        _properties = _properties.all()
        return _properties

    def update_property(self, payload):
        _property = self.session.query(Property).get({"id": payload["id"]})
        if (_property == None):
            app.logger.error('Property not found')
        self.session.query(Property).filter(Property.id == payload["id"]).update(payload)
        self.session.commit()
        return _property

    def delete_property(self, payload):
        _property = self.session.query(Property).get({"id": payload["id"]})
        if (_property == None):
            app.logger.error('Property not found')
        self.session.delete(_property)
        self.session.commit()
        return _property

    def get_all_property(self):
        # TO DO : changes corresponding to search property API
        return self.db_session.query(Property).all()
    
    
    
