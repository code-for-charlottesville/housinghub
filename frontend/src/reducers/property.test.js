import property from "./property";

describe("reducers/appState", () => {
  beforeEach(() => {
    Date.now = jest.fn(() => Date.parse("2020-06-01"));
  });


  let testTable = [
    {
      name: "set new prop",
      action: {
        type: "SET_NEWPROP_SUCCESS",
        id: "test id",
      },
    },
    {
      name: "set new prop failure",
      action: {
        type: "SET_NEWPROP_FAILURE",
        error: "new prop error",
      },
    },
    {
      name: "add property field (doesnt already exist)",
      action: {
        type: "SET_ADD_PROPERTY_FIELD",
        fieldName: "test field",
        newValue: "new test value",
      },
    },
    {
      name: "add property field (already exists)",
      action: {
        type: "SET_ADD_PROPERTY_FIELD",
        fieldName: "property_name",
        newValue: "new prop test",
      },
    },
    {
      name: "add property field (is array and already exists)",
      action: {
        type: "SET_ADD_PROPERTY_FIELD",
        fieldName: "where_listed",
        newValue: "new prop test",
      },
    },
  ];
  testTable.forEach((t) => {
    it(t.name, () => {
      expect(property(undefined, t.action)).toMatchSnapshot();
    });
  });
});
