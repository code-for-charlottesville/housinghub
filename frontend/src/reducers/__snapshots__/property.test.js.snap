// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers/appState add property field (already exists) 1`] = `
Object {
  "addProperty": Object {
    "fields": Object {
      "address": "",
      "allow_criminal_records": false,
      "application_fee": 0,
      "background_screening_company": "",
      "bathrooms": 0,
      "bedrooms": 0,
      "bus_line": false,
      "contact_method": Array [],
      "credit_screening_company": "",
      "date_first_available": "2020-06-18",
      "deposit": 0,
      "elevator": false,
      "floor": 0,
      "has_basement": false,
      "housing_type": "",
      "is_available": false,
      "landlord_id": null,
      "last_contact_date": "2020-06-18",
      "last_contacted_by": null,
      "last_month_rent_required": false,
      "listing_date": "2020-06-18",
      "monthly_rent": 0,
      "navigator_id": null,
      "potential_month_available": 0,
      "property_name": "new prop test",
      "school_district": "",
      "shared_bathrooms": 0,
      "unit_apt_no": "",
      "voucher_type_accepted": Array [],
      "voucher_type_not_accepted": Array [],
      "wheelchair_accessibility": false,
      "where_listed": Array [],
      "year_available": 2020,
      "zip_code": "",
    },
  },
  "error": "",
}
`;

exports[`reducers/appState add property field (doesnt already exist) 1`] = `
Object {
  "addProperty": Object {
    "fields": Object {
      "address": "",
      "allow_criminal_records": false,
      "application_fee": 0,
      "background_screening_company": "",
      "bathrooms": 0,
      "bedrooms": 0,
      "bus_line": false,
      "contact_method": Array [],
      "credit_screening_company": "",
      "date_first_available": "2020-06-18",
      "deposit": 0,
      "elevator": false,
      "floor": 0,
      "has_basement": false,
      "housing_type": "",
      "is_available": false,
      "landlord_id": null,
      "last_contact_date": "2020-06-18",
      "last_contacted_by": null,
      "last_month_rent_required": false,
      "listing_date": "2020-06-18",
      "monthly_rent": 0,
      "navigator_id": null,
      "potential_month_available": 0,
      "property_name": "",
      "school_district": "",
      "shared_bathrooms": 0,
      "test field": "new test value",
      "unit_apt_no": "",
      "voucher_type_accepted": Array [],
      "voucher_type_not_accepted": Array [],
      "wheelchair_accessibility": false,
      "where_listed": Array [],
      "year_available": 2020,
      "zip_code": "",
    },
  },
  "error": "",
}
`;

exports[`reducers/appState add property field (is array and already exists) 1`] = `
Object {
  "addProperty": Object {
    "fields": Object {
      "address": "",
      "allow_criminal_records": false,
      "application_fee": 0,
      "background_screening_company": "",
      "bathrooms": 0,
      "bedrooms": 0,
      "bus_line": false,
      "contact_method": Array [],
      "credit_screening_company": "",
      "date_first_available": "2020-06-18",
      "deposit": 0,
      "elevator": false,
      "floor": 0,
      "has_basement": false,
      "housing_type": "",
      "is_available": false,
      "landlord_id": null,
      "last_contact_date": "2020-06-18",
      "last_contacted_by": null,
      "last_month_rent_required": false,
      "listing_date": "2020-06-18",
      "monthly_rent": 0,
      "navigator_id": null,
      "potential_month_available": 0,
      "property_name": "",
      "school_district": "",
      "shared_bathrooms": 0,
      "unit_apt_no": "",
      "voucher_type_accepted": Array [],
      "voucher_type_not_accepted": Array [],
      "wheelchair_accessibility": false,
      "where_listed": Array [
        "new prop test",
      ],
      "year_available": 2020,
      "zip_code": "",
    },
  },
  "error": "",
}
`;

exports[`reducers/appState set new prop 1`] = `
Object {
  "addProperty": Object {
    "fields": Object {
      "address": "",
      "allow_criminal_records": false,
      "application_fee": 0,
      "background_screening_company": "",
      "bathrooms": 0,
      "bedrooms": 0,
      "bus_line": false,
      "contact_method": Array [],
      "credit_screening_company": "",
      "date_first_available": "2020-06-18",
      "deposit": 0,
      "elevator": false,
      "floor": 0,
      "has_basement": false,
      "housing_type": "",
      "is_available": false,
      "landlord_id": null,
      "last_contact_date": "2020-06-18",
      "last_contacted_by": null,
      "last_month_rent_required": false,
      "listing_date": "2020-06-18",
      "monthly_rent": 0,
      "navigator_id": null,
      "potential_month_available": 0,
      "property_name": "",
      "school_district": "",
      "shared_bathrooms": 0,
      "unit_apt_no": "",
      "voucher_type_accepted": Array [],
      "voucher_type_not_accepted": Array [],
      "wheelchair_accessibility": false,
      "where_listed": Array [],
      "year_available": 2020,
      "zip_code": "",
    },
  },
  "error": "",
  "id": "test id",
}
`;

exports[`reducers/appState set new prop failure 1`] = `
Object {
  "addProperty": Object {
    "fields": Object {
      "address": "",
      "allow_criminal_records": false,
      "application_fee": 0,
      "background_screening_company": "",
      "bathrooms": 0,
      "bedrooms": 0,
      "bus_line": false,
      "contact_method": Array [],
      "credit_screening_company": "",
      "date_first_available": "2020-06-18",
      "deposit": 0,
      "elevator": false,
      "floor": 0,
      "has_basement": false,
      "housing_type": "",
      "is_available": false,
      "landlord_id": null,
      "last_contact_date": "2020-06-18",
      "last_contacted_by": null,
      "last_month_rent_required": false,
      "listing_date": "2020-06-18",
      "monthly_rent": 0,
      "navigator_id": null,
      "potential_month_available": 0,
      "property_name": "",
      "school_district": "",
      "shared_bathrooms": 0,
      "unit_apt_no": "",
      "voucher_type_accepted": Array [],
      "voucher_type_not_accepted": Array [],
      "wheelchair_accessibility": false,
      "where_listed": Array [],
      "year_available": 2020,
      "zip_code": "",
    },
  },
  "error": "new prop error",
}
`;
